### DISCUSSION TITLE
Get selected nodes

### DISCUSSION THREAD

**Discussion by dominikspiertz** [xyflow/xyflow#2237] [Q&A] [answered]
Hello dear folks,
we are currently updating react flow from version 9 to 10. We have been using a lot of customization before and created a structure that would be best described as a controlled flow. We have our own redux store that creates elements (nodes and edges). It is almost completely possible for us to use react-flow with the old api's, which is a great experience for us. The only thing that is not working for us anymore is the handling of a node being selected (clicked) and unselected. In our code we expect to know which node(s) have been selected by the user, by clicking on them (We use custom nodes). If a user clicks on a node in our application, the selection seems to work as before, since our custom styling still works. It listens to the `selected` class to be attached to the node and adds a border. While this works on a styling level, I did not found a solution to get the selected node(s) from the store. We have been using `useStoreState(state.selectedElements)` before to receive the selected elements. If I read the migration guide correctly, this is now set by `selected: true` inside the node.

Besides that, I figured out that I can imperatively select nodes by using `useStore(state.addSelectedNodes)` where we used before `useStoreActions(actions.setSelectedElements)`. This seems to be the replacement function to select nodes. The question here is, how do we retrieve this state? By setting `addSelectedNodes`, we are updating the store for react-flow, rather than managing the selection on our own, which would resolve in a dispatch to our own store with a new node. Even If I click in the ui, the selected node is changed. If this is possible to set, but not to retrieve, the ability to do so, doesn't serve the purpose of having some part of 'controlled' or 'uncontrolled' behaviour. I have used `const { getNodes } = useReactFlow()` to get updates about the current nodes, but I guess it just returns my `controlled` nodes rather than the flow state of `nodes` which can be selected. I guess there has to be a way to also retrieve the selected node(s), since it can be affected by my actions in the code or by the user in the frontend.

Any help is appreciated, thanks.

**Comment by moklick** [comment, marked as answer]
Hey @dominikspiertz 

I am happy to hear that the migration from 9 to 10 doesn't feel too bad ;)

To your questions:

### Receive selected state:

If you are working with a controlled flow I assume that you are using the `onNodesChange` handler. For selecting nodes it works like this:

1. user selects node (or multiple ones)
2. internal selection change event gets fired 
3. `onNodesChange` updates your local `nodes` state / your store state
4. `selected` attribute is set to `true` for the selected node

### Set selected nodes:

In order to select a node programatically you need to set `selected: true` and pass the altered nodes to the React Flow component. There is no need to use `addSelectedNodes` in user land and you also don't need to use `getNodes` or something like that, if you are working with a controlled flow. You should always use your Redux store to receive the nodes and Redux actions to alter them. 

We recently added a [guide about how to use React Flow with a state management library](https://reactflow.dev/docs/guides/state-management/). That could be helpful to get a better idea of how to connect Redux with the new change handlers.

**Reply by dominikspiertz** [reply]
Hey @moklick thank you for the quick response and explanation. I just have a few questions to clarify. The way a controlled flow should behave now is clear to me. It seems that we have to go with the store updates to mark nodes as selected if we want it like that. In our previous version we have not been using `onNodesChange`. You could say the part about selecting / unselecting nodes was uncontrolled and we just got the `selectedElements` from react-flow and its store. It is a little inconvenient for the moment, that we need to adjust our data structure but given the idea of controlled and uncontrolled components just like in react, it does make sense to opt-in to either one of the two possibilities. I think our custom hooks will also be thankful for that, since the typing doesn't have to deal with the react-flow types and our custom types combined.

Could you clarify when it would make sense to call actions like `addSelectedNodes`. Is this api designed for the case if you do not have a controlled flow and you want to select nodes imperatively?

**Reply by moklick** [reply]
1. The problem with the previous version was that the internal nodes state and your local state were out of sync as soon as you started to drag or select a node for example. v10 solves this and it's now easier to understand what is going on with your nodes and edges.
2. Calling internal actions was always a hack. The new API should make it obsolete to call any internal actions. You are in control of the nodes and edges and you should only update those with your actions/state setter.

**Reply by dominikspiertz** [reply]
Agree, thanks for your time and your answer
